src/modules/roc_sndio/target_sox/roc_sndio/sox_backend.cpp

```diff
@@ -252,8 +268,11 @@ ISink* SoxBackend::open_sink(core::IAllocator& allocator,
         return NULL;
     }
 
-    if (!sink->open(driver, output)) {
-        return NULL;
+    while (!sink->open(driver, output)) {
+        driver = get_fallback_driver(driver);
+        if (!driver) {
+            return NULL;
+        }
     }
```

Comment:
This code will try to switch driver even if the user did specify the driver explicitly.

E.g. the user specified `pulse://default`, pulse is not running, and here we'll try `alsa://default`. This is not what the user asked. We should only try multiple drivers if the user didn't specify the driver at all.

A few lines above we call `select_defaults()`, which selects default driver in case if no driver was specified. But with our new approach there is no more such thing as a single default driver. So instead of selecting one default driver there, we should rework that piece and try drivers in a loop, until open succeeds, as you did it here.

Also, when we're trying multiple drivers, we should log the reason. Something like "failed to open sink FOO using driver BAR, trying next driver". Or some other message, the idea is to explain the user why we're doing what we're doing.